### DISCUSSION TITLE
Refactor pre-processor syntax

### DISCUSSION THREAD

**Discussion by anbraten** [woodpecker-ci/woodpecker#2473] [Request a Feature]
We currently use bash var syntax for pre-processing in our yaml config. So a var would be written by using `${CI_REPO_DEFAULT_BRANCH}`. Its using <https://github.com/drone/envsubst> for this. This sometimes conflicts with users trying to write variables with normal bash syntax. To escape this the user is needed to write: `$${CI_REPO_DEFAULT_BRANCH}`.

Instead maybe changing to some other syntax like go-template / handlebar syntax with `{{ CI_REPO_DEFAULT_BRANCH }}` for the pre-processor could avoid confusion.

**Comment by qwerty287** [comment]
We should still provide all the features supported by https://github.com/drone/envsubst (https://github.com/drone/envsubst#supported-functions), this shouldn't be hard to implement using something like Go templates though.

**Comment by qwerty287** [comment]
Related: https://github.com/woodpecker-ci/woodpecker/issues/1258

**Comment by anbraten** [comment]
As we can not drop the syntax directly we might need a roadmap like this:
- discuss a new pre-processor syntax
- add this syntax as early access to the ci
- improve the design
- deprecate old syntax, so users can migrate from `$$` to `$`
- remove old syntax and run it as "normal" bash

**Reply by qwerty287** [reply]
> discuss a new pre-processor syntax

The best would probably to just use go templates. They have a specific syntax so it's much harder to mis-interpret them, you can easily provide functions so you can use e.g. string replacing or prefix stripping, we wouldn't have to change much as it works pretty similar to the current approach, and it's stdlib.

**Reply by anbraten** [reply]
Yes I think so as well. using go template is probably the best.

**Reply by xoxys** [reply]
What is the real benefit of it? The current syntax is well documented and easy to escape. Not against such a change, but is yet another breaking change (especially to the workflow syntax) really worth it?

**Reply by anbraten** [reply]
Did some testing in #3341, but actually think its not 100% needed as the current logic skips vars if they are not set (see the tests I've added) and leaves sth like `${HELLO_WORLD}` untouched in case `HELLO_WORLD` is not set. 

It would be possible to support both ways, but not sure about that. 

One thing I really liked when using the go-template is writing sth simple like this:

```yaml
steps:
  - name: test
    image: my-plugin
    settings:
      dry-run: {{if eq env.CI_PIPELINE_EVENT "pull_request"}}true{{else}}false{{end}}
```

**Reply by xoxys** [reply]
That's what I would count as a real benefit :+1: thanks.

**Reply by qwerty287** [reply]
@anbraten 

> Did some testing in #3341, but actually think its not 100% needed as the current logic skips vars if they are not set (see the tests I've added)

Are you sure this is tested? Can't see the testcase there and there were some bugs some time ago due to this issue:
```yaml
steps:
  - name: x
    image: alpine:${MISSING}
```

gets into

```yaml
steps:
  - name: x
    image: alpine:
```

which would be invalid yaml what was the reason for these bug reports.

**Reply by 6543** [reply]
I think having a different syntax is ok

But I will veto to make the configuration scriptable as long as I'm a maintainer in fafour of a more generic solution!!!
-> #2254 

As I see there is quite a need and push towords something scriptable for configuration I will focus on that generic solution as one of my next priorities

**Reply by xoxys** [reply]
I dont get your point. This discussion was about go templates instead of envsubst-like syntax, no? No idea what you mean by "scriptable" and why you jump into the discussion with a negative tone.

**Reply by lafriks** [reply]
I completely agree with @6543, I also agree that making yaml file scriptable is not the way to go, we could support plugins that would allow to use scripting languages to generate yaml files etc but adding go syntax in yaml would be a mess imho

**Reply by xoxys** [reply]
But having a bash-like syntax (and limited substitution functions) in yaml is better then? This is also just another "hack" to get information templated into the pipeline config in a dynamic way...

**Reply by 6543** [reply]
I agree and that's why i wrote to switch to handlebare like syntax for simple variable replacement, we can do ...

but the example provide was about having the whole template **engine** in place

**Reply by anbraten** [reply]
Just to clarify the initial proposal: We currently use some templating that matches the one from bash (but is not bash => https://github.com/drone/envsubst). The UX issue I wanted to highlight here is that inexperienced users often expect `${...}` to be some bash stuff executed in the backend like docker and therefore mixes up where which variables are available. Using some other syntax for variables (let it be `%%THIS_IS_MY_VAR%%` or whatever) would clarify this. Mainly used the go / handlebars syntax as it is pretty common for such things.

**Reply by lafriks** [reply]
Looking at others gitlab ci uses same as we standard env `${ ... }` syntax, github ci `${{ ... }}` syntax, to be honest I would prefer to keep as we have now, maybe change to github syntax as that would at least indicate that this is not full go template syntax but something more specific (this would probably mean forking envsubst lib to support this

**Reply by qwerty287** [reply]
Maybe it could also be helpful to not run the templating on the whole yaml file, but only on some specific fields like env vars, settings images etc. Commands could be left out and then bash (or which shell your container has) parses it.

**Reply by zc-devs** [reply]
> Commands could be left out and then shell parses it

I second this.
#3982, #3983.

---

Edit 1
> specific fields like env vars

Do we need substitution there? Cannot imagine a use case. Even if I did, it would pretty useless, because syntax would be very simple, it would not support full-Bash substitution, I assume. We can do substitution in `commands` whatever we want.

**Reply by qwerty287** [reply]
Yes, but for plugin settings, we need some more advanced features (example: removing the `v` prefix of a tag) which is not directly done by the plugin. And if we do it for settings, we can also do it for env vars.

**Reply by zc-devs** [reply]
> if we do it for settings, we can also do it for env vars

If we do it for env vars, we can also do it for commands, then 😄 
Plugins is a different story. There is no envs now (you know).

I'm not pushing, just consider to leave out `environment` as well.

**Comment by 6543** [comment]
we should not make sinple config turing complete!!!

https://www.youtube.com/watch?v=yYaLQ3LazYM

https://www.grc.com/sn/sn-919.htm (GIF hack where pdf rendering engine was turing complete and so hackers could break out)
